### ISSUE TITLE
Smithy2OpenAPI return multiple responses codes (besides errors)

### ISSUE THREAD

**Issue by NukaCody** [awslabs/smithy#1308]
We have a configuration resource

```smithy
resource Configuration {
   put: PutConfiguration,
   read: GetConfiguration,
   delete: DeleteConfiguration,
}

@idempotent
@http(method: "PUT", uri: "/configuration")
operation PutConfiguration {
    input: PutConfigurationInput,
    output: PutConfigurationOutput,
    errors: [InvalidCommand, NotFound, Forbidden],
}

@readonly
@http(method: "GET", uri: "/configuration")
operation GetConfiguration {
    output: GetConfigurationOutput,
    errors: [InvalidCommand, NotFound, Forbidden]
}

@idempotent
@http(method: "DELETE", uri: "/configuration")
operation DeleteConfiguration {
    errors: [InvalidCommand, Forbidden]
}
```

Which generates a good OpenAPI spec

```yaml
paths:
  /configuration:
    (...GET and DELETE rm for brevitity)
    put:
      operationId: PutConfiguration
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/PutConfigurationRequestContent'
        required: true
      responses:
        '200':
          description: PutConfiguration 200 response
          headers:
            Content-Type:
              description: Value MUST be "application/cdni;ptype=ci-trigger-status"
              schema:
                type: string
                description: Value MUST be "application/cdni;ptype=ci-trigger-status"
              required: true
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TriggerStatus'
        '400':
          description: InvalidCommand 400 response
        '403':
          description: Forbidden 403 response
        '404':
          description: NotFound 404 response
```

Just realized that it renders content as `application/json` but we need it to render a custom content type "application/cdni;ptype=MI.HostIndex" to not deviate from the RFCs. Not sure if there is anything we can do about that because we are also leveraging `@restJson1` but this may be a separate issue.

To the main point, what we want to do is add more response codes. 

```
      responses:
        '102':
           description: Some generic description that we are currently processing the request
        '200':
          description: PutConfiguration 200 response
          headers:
            Content-Type:
              description: Value MUST be "application/cdni;ptype=ci-trigger-status"
              schema:
                type: string
                description: Value MUST be "application/cdni;ptype=ci-trigger-status"
              required: true
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TriggerStatus'
        '201':
           description: Alternative to 200 but with Location header
        '400':
          description: InvalidCommand 400 response
        '403':
          description: Forbidden 403 response
        '404':
          description: NotFound 404 response
```

We initially tried

```smithy
@output
structure PutConfigurationOutput {
    @httpResponseCode
    code: Integer,

    /// Value MUST be "application/cdni;ptype=ci-trigger-status"
    @required
    @httpHeader("Content-Type")
    applicationCdni: String,

    /// The status of any configuration updates
    @httpPayload
    content: TriggerStatus
}
```

Which works from a code generation perspective, but Smithy2OpenAPI does not know what response codes to generate so it defaults to 200 which is what was defined at `@http(method: "PUT", uri: "/configuration")`

We then tried to move one up by defining multiple operations on the configuration resource.

```smithy
resource Configuration {
   put: PutConfiguration,
   read: GetConfiguration,
   delete: DeleteConfiguration,
   operations: [CurrentlyProcessing]
}

@idempotent
@http(method: "PUT", uri: "/configuration", code: 102)
operation CurrentlyProcessing {
}
```

However, since this conflicts with the normal PutConfiguration URI, this is not allowed and Smithy throws an error. Smithy additionally throws another error (although danger in this case) because it expects an HTTP code in the 2xx range, even though 102 is a valid HTTP response code.

```bash
DANGER: #CurrentlyProcessing (HttpResponseCodeSemantics)
     @ /model/main.smithy
     |
 135 | @http(method: "PUT", uri: "/configuration", code: 102)
     |       ^
     = Expected an `http` code in the 2xx range, but found 102
```

```bash
ERROR: #CurrentlyProcessing (HttpUriConflict)
     @ /model/main.smithy
     |
 136 | operation CurrentlyProcessing {
     | ^
     = Operation URI, `/configuration`, conflicts with other operation URIs in the same service: [`#PutConfiguration` (/configuration)]

```

As a last ditch effort, we tried errors inside of the existing PutConfiguration operation.

```smithy
@idempotent
@http(method: "PUT", uri: "/configuration")
operation PutConfiguration {
    input: PutConfigurationInput,
    output: PutConfigurationOutput,
    errors: [PutConfigurationOutputAlt, CurrentlyProcessing, InvalidCommand, NotFound, Forbidden],
}

@error("server")
@httpError(201)
structure PutConfigurationOutputAlt {
}

@error("server")
@httpError(102)
structure CurrentlyProcessing {
}
```

However this fails because `@httpError()` expects a 5xx range for "server" and 4xx for client. The error tag is also mandatory to specify inside of errors.

I think the ideal scenario would be to allow us to define multiple operations for the same URI although I'm not 100% sure what the downstream impact of that may be.

**Comment by mtdowling** [comment]
A couple things to keep in mind about Smithy:
1. Smithy provides some basic support for customizing how messages are sent over HTTP. The lens in which HTTP is modeled in Smithy still fits within the simple request/response/errors pattern though. Deviating from this design tenet is possible to an extent, but there are tradeoffs.
2. Not every feature of OpenAPI is expressable in Smithy. While Smithy can be converted to OpenAPI, Smithy isn't just a thin translation layer on top of OpenAPI. Some of the gaps between what Smithy and OpenAPI express can be patched in using the `jsonAdd` property of Smithy's `openapi` plugin. If you need to do something drastically different than what Smithy supports and are looking for an extremely specific OpenAPI output that can't be easily patched using `jsonAdd`, then Smithy might not be the right tool.

> we need it to render a custom content type "application/cdni;ptype=MI.HostIndex" to not deviate from the RFCs

You should probably look at using `jsonAdd` to unblock yourself if this is necessary.

The restJson1 protocol from AWS doesn't support changing the content type of responses like this. You can change the content-type by using the `@httpPayload` trait that targets a blob/string marked with the `@mediaType` trait, but not when targeting a structure. Note that even if this were possible, "application/cdni;ptype=MI.HostIndex" would have to be JSON to match the rest of the protocol, and clients would completely ignore this deviation.

> To the main point, what we want to do is add more response codes. [102, 201]

There are two ways to do this. One is to use `@httpResponseCode`, the other is to model those as errors (a third is to avoid using these status codes, but not sure if that's an option for you). I would use [`@httpResponseCode`](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpresponsecode-trait) to differentiate betwee 200 and 201 (and I suppose it would make sense for 102 as well, but I don't know much about 102). Using errors has other protocol implications like needing to serialize the `X-Amzn-Errortype` header.

The `@httpResponseCode` trait on an output member gives an operation a dynamic status code. [Amazon Glacier uses this trait](https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/sdk-codegen/aws-models/glacier.json#L1256) to return 200 or 201 responses.

```
... other stuff...
structure PutConfigurationOutput {
    ...other members...
    @httpResponseCode
    status: Integer
}
```

When serializing the response, change the status code of the response by setting the `status` member to the status code 102 or 201. Each status code doesn't get a dedicated structure though; you will need to ensure that the 200 response structure is compatible with any other successful response you return though. That means that if 102 returns nothing, then all output members need to be optional. If 201 returns nothing + a Location header, then all output members need to be optional, and you need to model a member bound to the `Location` header using `@httpHeader("Location")`.

The 102 response and 201 response won't show up in OpenAPI if you use this approach, but even OpenAPI isn't required to explicitly model every possible status code that can be returned. HTTP clients are expected to be able to handle unmodeled status codes. For example, it's often not necessary to model status codes like 5xx unless there are explicit schemas in these responses you need clients to parse.

If you need these status codes to appear though, you can use `jsonAdd`.

> Smithy additionally throws another error (although danger in this case) because it expects an HTTP code in the 2xx range, even though 102 is a valid HTTP response code.

You can suppress this. I think it's appropriate to emit this danger event because 102 is extremely rare and should be used with care.

> I think the ideal scenario would be to allow us to define multiple operations for the same URI although I'm not 100% sure what the downstream impact of that may be.

You can define multiple operations to the same URI, but just not to the same URI _and_ HTTP method. That would result in ambiguous routing. Even still, I don't think that would be a solution for you. You'd be asking clients to pick how the service responds by choosing to use `PutConfiguration` or `PutConfigurationAlt`, but the client doesn't know if the server will return a 200, 201, or 102, and it shouldn't be able to dictate the server response.
